src/client/datascience/interactive-window/nativeInteractiveWindow.ts

```diff
@@ -122,6 +128,8 @@ export class NativeInteractiveWindow implements IInteractiveWindowLoadable {
     private kernel: IKernel | undefined;
     private kernelLoadPromise: Promise<void> | undefined;
     private initialControllerSelected: Deferred<void>;
+    private _readyPromise: Promise<NotebookDocument>;
+    private notebookDocument: NotebookDocument | undefined;
```

Comment:
Seems like you should't catch the document here. Just use the readyPromise everywhere. Or maybe call it the 'documentPromise' or somehting.

Reply:
We need this in order to support synchronously finding a matching interactive window based on this.notebookDocument.uri, in cases where having an undefined notebookUri doesn't impact correctness, e.g. getting the active interactive window. If there's no notebookDocument in that case, by definition the UI isn't up, and we read activeWindow to set context keys so awaiting on the ready promise in such scenarios is unnecessary.

Reply:
Let me make sure I understand. The notebookDocument.URI is necessary synchonously so that the context keys can update without having to await?

Reply:
Wow, sorry, I meant we read the activeWindow to handle commands (in the command listener). No context keys involved. But the commands don't make sense without the UI being up so there shouldn't be a difference if we read notebookUri without awaiting on readyPromise versus awaiting and getting notebookDocument.uri.